### ISSUE TITLE
AWSPinpointVersionString Missing

### ISSUE THREAD

**Issue by kwallace-abvio** [aws/aws-sdk-ios#1050]
**Describe the bug**
When I link with the AWS SDK for iOS 2.6.30 CocoaPod, AWSPinpointVersionString is missing.  I see it referenced in the source code but not defined/initialized.

**To Reproduce**
Include the AWSPinpoint CocoaPod and link with it.

**Which AWS service(s) are affected?**
AWSPinpoint

**Expected behavior**
My app should link.

**Screenshots**

Undefined symbols for architecture arm64:
  "_AWSPinpointVersionString", referenced from:
      -[AWSPinpointEventRecorder buildEventPayload:] in libAWSPinpoint.a(AWSPinpointEventRecorder.o)
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)

**Comment by kvasukib** [comment]
@kwallace-abvio Sorry for the inconvenience caused. This issue has been fixed in 2.6.31. Can you consume the latest to see if the issue has been fixed?

**Comment by kwallace-abvio** [comment]
It doesn't appear to be fixed to me.  I did a pod update/install and I now have version 2.6.31 of all the AWS pods.  The problem still occurs.  I had previously done a git clone of aws-sdk-ios and I just did a git pull to get the latest.  I did the following find:

% find aws-sdk-ios -type f -exec grep AWSPinpointVersionString {} /dev/null \;
aws-sdk-ios/AWSPinpoint/AWSPinpoint.h:FOUNDATION_EXPORT const unsigned char AWSPinpointVersionString[];
aws-sdk-ios/AWSPinpoint/AWSPinpointEventRecorder.m:    serviceEvent.clientSdkVersion = [NSString stringWithFormat:@"%s", AWSPinpointVersionString];
% 

The FOUNDATION_EXPORT in AWSPinPoint.h isn't sufficient.  In what file do you expect AWSPinpointVersionString to be defined?

**Comment by TriggerJock** [comment]
I have exactly this problem as well.  Unfortunately I was right in the middle of re-building a project and for the longest time was certain I was doing something wrong.

Any resolution on this?  Failing that, what version should I fall back to instead of 2.6.31 for all my AWS pods until a fix comes out?

Thanks.

**Comment by muellerfr** [comment]
Thank you for reporting this issue. We are looking into the problem.

**Comment by Gvdpoel** [comment]
Having the exact same problem. The work around I used was to set our Podfile to use version 2.6.27.

pod 'AWSCore', '= 2.6.27'
pod 'AWSMobileAnalytics', '= 2.6.27'
pod 'AWSMobileClient', '= 2.6.27'
pod 'AWSAuthCore', '= 2.6.27'
pod 'AWSPinpoint', '= 2.6.27'
pod 'AWSSNS', '= 2.6.27'

**Comment by kvasukib** [comment]
@kwallace-abvio @Gvdpoel @TriggerJock Can you share the reproduction steps? So far, I tried the following and I didn't notice any error.

1) cat Podfile                                                                                             
platform :ios, '9.0'
target :'PinpointTester' do
  use_frameworks!

    pod 'AWSPinpoint', '~> 2.6.31'
    pod 'AWSMobileClient', '~> 2.6.31'

    # other pods

end

2) pod install
3) Open project in XCode (9.2), did a build and build succeeded.

**Comment by TriggerJock** [comment]
Here's my Podfile:

```
# Uncomment this line to define a global platform for your project
platform :ios, '9.3'

source 'https://github.com/CocoaPods/Specs.git'
source 'https://github.com/deltaDNA/CocoaPods.git'

def shared_pods

pod 'AWSiOSSDKv2', '~> 2.4'
pod 'AWSCognito', '~> 2.4'
pod 'AWSLambda'
pod 'AWSMobileAnalytics', '~> 2.4'
pod 'AWSMobileClient'
pod 'FBSDKCoreKit', '~> 4.22'
pod 'FBSDKLoginKit', '~> 4.22'
pod 'FBSDKShareKit', '~> 4.22'
pod 'SSZipArchive', '2.0.8'
pod 'VungleSDK-iOS'
pod 'OpenSSL-Universal'
pod 'DeltaDNA', '4.5.0'

end

target 'Casino-mobile' do
       platform :ios, '9.3'
       shared_pods
end

target 'CompCity-mobile' do
       platform :ios, '9.3'
       shared_pods
end

target 'Binions' do
       platform :ios, '9.3'
       shared_pods
end

target 'Herbst' do
       platform :ios, '9.3'
       shared_pods
end

target 'FourQueens' do
       platform :ios, '9.3'
       shared_pods
end

```

The '~> 2.4' does install 2.6.31.  I added AWSMobileClient while trying to debug this.  I don't use AWSMobileClient (My project is 4 years old).

I'm going to try 2.6.27 to see if that works.

**Comment by TriggerJock** [comment]
Switching to 2.6.27 compiled, linked, and ran successfully.

Here's the changed portion of my Podfile:

```
def shared_pods

pod 'AWSiOSSDKv2', '2.6.27'
pod 'AWSCognito', '2.6.27'
pod 'AWSLambda', '2.6.27'
pod 'AWSMobileAnalytics', '2.6.27'
pod 'FBSDKCoreKit', '~> 4.22'
pod 'FBSDKLoginKit', '~> 4.22'
pod 'FBSDKShareKit', '~> 4.22'
pod 'SSZipArchive', '2.0.8'
pod 'VungleSDK-iOS'
pod 'OpenSSL-Universal'
pod 'DeltaDNA', '4.5.0'

end

```

**Comment by kvasukib** [comment]
@TriggerJock Is there a reason why you are using `pod 'AWSiOSSDKv2', '~> 2.4'` ? Instead, you could just declare the individual dependencies. For example, if you need `AWSPinpoint`, I would recommend using `pod 'AWSPinpoint', '~> 2.6'`.

Regardless, I am trying to reproduce the issue.

**Comment by TriggerJock** [comment]
@kvasukib No, not really.  It was probably the right choice years ago.

I've removed AWSiOSSDKv2 and got it working.  I also went back to the latest version, since I don't use Pinpoint, and it compiled and linked just fine.  Yay!

Here's my final pod settings:

```
def shared_pods

pod 'AWSDynamoDB'
pod 'AWSS3'
pod 'AWSSES'
pod 'AWSCognito'
pod 'AWSLambda'
pod 'AWSSNS'
pod 'AWSMobileAnalytics'
pod 'FBSDKCoreKit', '~> 4.22'
pod 'FBSDKLoginKit', '~> 4.22'
pod 'FBSDKShareKit', '~> 4.22'
pod 'SSZipArchive', '2.0.8'
pod 'VungleSDK-iOS'
pod 'OpenSSL-Universal'
pod 'DeltaDNA', '4.5.0'

end

```

**Comment by kwallace-abvio** [comment]
I'm not sure why this issue is so difficult to diagnose or address for whomever works on the SDK.  Simply looking at the source code, AWSPinpointVersionString is referenced in AWSPinpointEventRecorder.m.  There is a declaration of AWSPinpointVersionString  in AWSPinpoint.h, but AWSPinpointVersionString is not defined anywhere I can find.  I'm not sure what value AWSPinpointVersionString should take but my workaround is to simply add the following line of code somewhere in my source Objective-C code to satisfy the linker error.

`const unsigned char AWSPinpointVersionString[] = "2.6.31";`

At some point when I update my pods and get a duplicate symbol error, I'll remove the line.

**Comment by palpatim** [comment]
It looks like the issue is related to importing the library as a static library via CocoaPods. Since it's not a framework, the `AWSPinpointVersionString` is never populated. However, even importing it as a framework causes issues, since CocoaPods creates its own umbrella framework for each Pod, and populates it with an unexpected version value (See #1051). The solution proposed by @kwallace-abvio is not ideal since it actually causes errors attempting to compile the SDK, since Xcode/clang will automatically implement the `*VersionString` and `*VersionNumber` symbols as part of a framework build.

The fixes we elected to make are:
1. Deprecate all of the `*VersionNumber` and `*VersionString` symbols wherever they exist
1. Expose the `*SDKVersion` symbols that are already present in the service classes
1. Replace the use of `AWSPinpointVersionString ` symbol in `AWSPinpointEventRecorder.m` with the newly-exposed `AWSPinpointTargetingSDKVersion`.

It's not an ideal solution, since it doesn't really address the underlying issue of CocoaPods mirroring/shadowing/overwriting our own header declarations, but since these version strings aren't likely used much in practice outside, it feels like the right approach.

This should be fixed in the next release.

**Comment by palpatim** [comment]
This is released on 2.6.33. Please let me know if you see any issues with the new version.

**Comment by kwallace-abvio** [comment]
AWSPinpointVersionString no longer causes a CocoaPods linking issue as of 2.6.33.  Thanks!
